### DISCUSSION TITLE
ARM64 cross-compilation with CUDA backend

### DISCUSSION THREAD

**Discussion by george-pippos** [AdaptiveCpp/AdaptiveCpp#1665] [Ideas]
I have successfully built AdaptiveCpp with SSCP, the CUDA and OpenCL backends. This was painless but my end goal is to cross-compile for an arm64 nvidia platform (orin series). Essentially I want to be in a place where I can compile an application with the OpenCL backend for an x86 host and with the CUDA backend for an aarch64 host. Is this a supported use-case for this compiler?

**Comment by illuhad** [comment]
You will need to build the application twice. Once for x86, and once for aarch64. AdaptiveCpp can make kernels portable, but it cannot make the host binary with all dependencies it might have portable across architectures. It cannot recompile x86 system libraries that your application might depend on for you :)

For this you will need to build the AdaptiveCpp toolchain twice, once for x86 and aarch64.
For aarch64, in principle it would be like cross-compiling other applications as well. But I'm not sure to what extent the aarch64 CUDA toolkit is available on x86. You will also need an aarch64 LLVM, either by building it yourself or from some package.
AdaptiveCpp (at least its main compiler) is a JIT compiler, so LLVM would also be needed on the target.

Long story short: It might work, but I'm not sure if somebody has tried this already, and it's most certainly nothing that is regularly tested. If you can, it will be much easier to just build AdaptiveCpp on the target.

**Reply by george-pippos** [reply]
Looks like I'm out of luck.

I tried this with gcc but turns out this build calls clang under the hood as well and my sysroot / cross-compilation config wasn't being applied to the clang build.

Changing to clang for the build yielded more success. I set the sysroot and set the target. The project configured and started building but it seems the llvm used to link the compiler against is the x86 one for the libllvm-to-backend.so target. I've tried to set library / LD_LIBRARY paths but no go. I am guessing somewhere in the build this target is linked against whatever is set in `-DLLVM_DIR=...` during configuration. Whenever I have this set to the host's llvm, it yields the below. Setting this to the target's llvm yields `aarch64-binfmt-P: Could not open '/lib/ld-linux-aarch64.so.1': No such file or directory`.

```
[ 68%] Linking CXX shared library libllvm-to-backend.so
/usr/bin/aarch64-linux-gnu-ld: /usr/lib/llvm-16/lib/libLLVM-16.so.1: error adding symbols: file in wrong format
clang: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [src/compiler/llvm-to-backend/CMakeFiles/llvm-to-backend.dir/build.make:194: src/compiler/llvm-to-backend/libllvm-to-backend.so] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:539: src/compiler/llvm-to-backend/CMakeFiles/llvm-to-backend.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

Because I am doing this in a docker build which cross compiles both arm64 and amd64 binaries for our stack, I really wanted to be able to do also cross-compile adaptivecpp and then cross-compile the stack with the compiler. We may have to just rely on emulation for the arm64 build :sob:.

Cross-compiling a compiler to cross-compile a project which uses SYCL to JIT compile for different devices seems to be one step too far :cry:. I wanted the one compiler to rule them all but I think I may be attempting one too many levels of cross-compilation.

**Comment by george-pippos** [comment]
Ok I have done some more with this and tried what you said @illuhad. I did the below:

- Built AdaptiveCpp on x86 with the OpenCL backend
- Built AdaptiveCpp on ARM64 with the CUDA backend
- Attempted cross-compiling for ARM64 on x86

I initially used a toolchain file to set the paths and flags using a sysroot but not all of these are honoured during cross-compilation. I did some digging and found that the way SYCL is added to targets is not compatible with a cross-compilation workflow. `AdaptiveCpp/cmake/adaptivecpp-config.cmake:135` simply adds the AdaptiveCpp runtime libraries that are found when `find_package(AdaptiveCpp REQUIRED)` is run in the project's CMake configuration step to the target. This means that it will always be adding the host's runtime libraries to the target.

At this stage, I am going to recommend my workplace uses emulation / a combination of x86 cross-compilation and arm64 native compilation for compiling the components of the stack that use SYCL as opposed to persevering with / hacking together a cross-compilation with AdaptiveCpp.

I really love this project and think it is incredibly valuable to the community. Thank you for all your work.

So to conclude this saga: I'd like to request cross-compilation be added as a feature to really overtake Intel's implementation of SYCL especially given the industry is moving to ARM and the world we live in will be x86/ARM for a long time.
